examples/nginx-grafana-metabase/README.md

```diff
@@ -0,0 +1,162 @@
+  
+
+
+
+
+### Random Streams
+There is an important note about the design of random streams in Timeplus Proton mentioned in the [`RANDOM STREAM`](https://docs.timeplus.com/proton-create-stream#create-random-stream) documentation that is worth highlighting here:
+> [!NOTE]
+> The data of random stream is kept in memory during the query time. If you are not querying the random stream, there is no data generated or kept in memory.
+
+Essentially, each time you run a query like `SELECT remote_ip, rfc1413_ident, remote_user, date_time, http_verb, path, http_ver, status, size, referer, user_agent, malicious_request FROM nginx_access_log LIMIT 1;`, you'll get a different (random) result from the stream.
+
+For instance, running that query yields the following output:
+```sql
+SELECT
+  remote_ip, rfc1413_ident, remote_user, date_time, http_verb, path, http_ver, status, size, referer, user_agent, malicious_request
+FROM
+  nginx_access_log
+LIMIT 1
+
+Query id: d4b7f92c-bcd8-42ec-94ef-655eae86861b
+
+┌─remote_ip─────┬─rfc1413_ident─┬─remote_user─┬───────────────date_time─┬─http_verb─┬─path─────────┬─http_ver─┬─status─┬──size─┬─referer─┬─user_agent────┬─malicious_request─┐
+│ 181.12.95.134 │ -             │ -           │ 2024-05-02 00:00:00.000 │           │ /sitemap.xml │          │    302 │ 21345 │ -       │ Plenary/4.6.2 │                   │
+└───────────────┴───────────────┴─────────────┴─────────────────────────┴───────────┴──────────────┴──────────┴────────┴───────┴─────────┴───────────────┴───────────────────┘
+
+1 row in set. Elapsed: 0.012 sec.
+```
+
+Running it again, yields this output:
+```sql
+SELECT
+  remote_ip, rfc1413_ident, remote_user, date_time, http_verb, path, http_ver, status, size, referer, user_agent, malicious_request
+FROM
+  nginx_access_log
+LIMIT 1
+
+Query id: 11d071e4-6028-47da-826d-9db42176f8e5
+
+┌─remote_ip──────┬─rfc1413_ident─┬─remote_user─┬───────────────date_time─┬─http_verb─┬─path─────────┬─http_ver─┬─status─┬──size─┬─referer────────────┬─user_agent─────────────────────────────────────────┬─malicious_request─┐
+│ 204.25.223.210 │ -             │ -           │ 2024-05-15 00:00:00.000 │ POST      │ /favicon.ico │ HTTP/1.1 │    301 │ 32062 │ https://ayewo.com/ │ NetNewsWire (RSS Reader; https://netnewswire.com/) │                   │
+└────────────────┴───────────────┴─────────────┴─────────────────────────┴───────────┴──────────────┴──────────┴────────┴───────┴────────────────────┴────────────────────────────────────────────────────┴───────────────────┘
+
+1 row in set. Elapsed: 0.013 sec. 
+```
+
+Due to this design decision, we will need to create a snapshot from the random stream to serve as our test data. This snapshot will represent our access log data so we can reliably geolocate the latitude and longitude of each (randomly generated) IP address before we can correctly visualize them in Grafana.
+
+Afterwards, the article will use the random stream to illustrate streaming queries in Grafana.
+
```

Comment:
Look forwards to more content about the SQL and Grafana integation.

Reply:
Yes, of course.

I get occasional stability issues when using the Timeplus plugin with Grafana that requires me to restart the Docker container, e.g.:
```bash
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:42.119403134Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:46.375972748Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=892 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:46.376686483Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:47.136723585Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:47.137064412Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=900 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:51.393905513Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=907 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:51.393964919Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:52.146066063Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=915 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:52.146175035Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:56.411043054Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=924 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:56.411227803Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:18:57.167190451Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=931 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:18:57.167402103Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:01.42450791Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=939 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:01.42481641Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:02.181165871Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:02.181467678Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=947 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:06.440446079Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=955 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:06.440538286Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:07.201749037Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=963 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:07.201899997Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=grafana-apiserver t=2024-07-14T11:19:07.73712243Z level=info msg="[core] [Channel #245 SubChannel #246]grpc: addrConn.createTransport failed to connect to {Addr: \"unused\", ServerName: \"unused\", }. Err: connection error: desc = \"transport: error while dialing: dial unix /tmp/plugin3391027809: connect: connection refused\""
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:11.454859501Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=971 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:11.454996798Z level=error msg="Error running stream, re-establishing" path=350d7060-7796-4b8b-9729-ef1e787150d6 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
03_nginx-grafana-metabase-grafana-1     | logger=plugin.timeplus-proton-datasource t=2024-07-14T11:19:12.217220448Z level=error msg="plugin process exited" plugin=/var/lib/grafana/plugins/timeplus-proton-datasource/gpx_timeplus_proton_linux_amd64 id=979 error="exit status 2"
03_nginx-grafana-metabase-grafana-1     | logger=live.runstream t=2024-07-14T11:19:12.217427932Z level=error msg="Error running stream, re-establishing" path=937289dd-eee0-49b7-a175-e3f7d2fdf850 error="error running stream: rpc error: code = Unavailable desc = error reading from server: EOF" wait=5s
```